### PR TITLE
lots of code changes and new features

### PR DESCRIPTION
**Pull request by R-Abbasi**
Current changes in a view:

Capture1

- changes in `generateTrainingData.cpp`:
  - warning popup if nothing suitable to upload is chosen beforehand 
  - select a file and its data will be set into the fields 
    - fields are then grayed-out 
    - fields with no data from the file are capable of taking data
    - fields with data from file are read-only
    - fields get back to default settings if _Waveform From File_ radio button is unchecked 
    - fields show the data again if _Waveform From File_ radio button is checked again 
  - data of the chosen file from the _Devices & Waveforms Menu_ is set to the fields as soon as it's checked
     - fields get back to default settings if the file is unchecked 
     - fields get back to default settings if a device ID is selected or if neither of the radio buttons is checked 
     - choose a device ID of the current waveform, optionaly feed the fields and upload data
  - The data is sent to the url _https://postapi.aw4null.de/v1/uploadFile/_ if the `Send` button is clicked
    - a message pop ups informing the user of the result of the data upload 
    - data will also be written to a `.json` file to be able to observe the uploaded data (optional)
- modified `saves_popup.hpp` to `saves_popup.cpp` 
- extended the language maps (inside newly created header `languages.hpp`) to support the _generateTrainingData_ popup completely 
- modified `sendData.hpp` to upload data
- used an `enum class` for the key of the languages `std::map`s since string literal comparisons might not always be correct. 
- and more ...

PS: Didn't touch the part for which suggested a method for refactoring

### CODE REVIEW

File: src/apihandler.hpp [hunk 1]
```diff
@@ -1,142 +1,142 @@
...
+#include "jasonhandler.hpp"
+#include <curl/curl.h>
+#include <imgui.h>
+#include <nlohmann/json.hpp>
+#include <nlohmann/json_fwd.hpp>
+
+inline std::string send_to_api(nlohmann::json const &config,
+                               std::string const &file, std::string const &vin,
+                               std::string const &scantype) {
+  std::string api_message = "empty";
```
Comment (NP199) [inline src/apihandler.hpp:12]:
```suggestion
  std::string api_message{"empty"};

```

File: src/apihandler.hpp [hunk 2]
```diff
@@ -1,142 +1,142 @@
...
+
+  if (curl) {
+    std::string url =
+        load_json<std::string>(config, "api", "url") + "?" +
+        load_json<std::string>(config, "api", "input1") + "=" + vin + "&" +
+        load_json<std::string>(config, "api", "input2") + "=" + scantype;
+
+    // Hinzufügen der Daten als Formfelder
+    curl_mime *mime = curl_mime_init(curl);
+    curl_mimepart *part;
```
Comment (NP199) [inline src/apihandler.hpp:27]:
```suggestion
    curl_mimepart *part = curl_mime_addpart(mime);

```

File: src/apihandler.hpp [hunk 3]
```diff
@@ -1,142 +1,142 @@
...
+
+    struct curl_slist *header_list = NULL;
+    header_list = curl_slist_append(
+        header_list, load_json<std::string>(config, "api", "header1").c_str());
+    header_list = curl_slist_append(
+        header_list, load_json<std::string>(config, "api", "header2").c_str());
+
+    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, header_list);
+    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
+    // curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
```
Comment (NP199) [inline src/apihandler.hpp:43]: comment needed?

File: src/apihandler.hpp [hunk 4]
```diff
@@ -1,142 +1,142 @@
...
+#include <imgui.h>
+#include <nlohmann/json.hpp>
+#include <nlohmann/json_fwd.hpp>
+
+inline std::string send_to_api(nlohmann::json const &config,
+                               std::string const &file, std::string const &vin,
+                               std::string const &scantype) {
+  std::string api_message = "empty";
+  CURL *curl;
+  CURLcode res;
```
Comment (NP199) [inline src/apihandler.hpp:14]: Is it needed here?
res is used in line 48

File: src/apihandler.hpp [hunk 5]
```diff
@@ -1,142 +1,142 @@
...
+  }
+
+  // Aufräumen von cURL
+
+  curl_global_cleanup();
+
+  return api_message;
+}
+
+inline std::string send_to_api(nlohmann::json const &config,
```
Comment (NP199) [inline src/apihandler.hpp:74]: The suggestion as from the suggestion above.

File: src/create_training_data.hpp
Comment (NP199) [inline src/create_training_data.hpp:1]: What are the changes on this file?

File: src/jasonhandler.hpp
Comment (NP199) [inline src/jasonhandler.hpp:1]: What are the changes to this file?

File: src/languages.hpp [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    {Key::AnalyzeData, "Daten\nAnalysieren"},
+    {Key::Crt_trng_data, "Erstellen\nSie\nTrainingsdaten"},
+    {Key::Back, "Zurück"},
+    {Key::Start, "Start"},
+    {Key::Stop, "Stoppen"},
+    {Key::Continue, "Weitermachen"},
+    {Key::Devices_found, "Geräte Gefunden"},
+    {Key::Usr_curnt_wave, "Aktuelle Wellenform Des Benutzers"},
+    {Key::Wv_from_file, "Wellenform Aus Datei"},
+    {Key::WvForms_warning, "Warnung vor Wellenformen! "},
```
Comment (NP199) [inline src/languages.hpp:154]:
```suggestion
    {Key::WvForms_warning, "Warnung vor Wellenformen!"},

```

File: src/languages.hpp [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    {Key::No_wave_made, "Es wurden keine Wellenformen erstellt! "},
```
Comment (NP199) [inline src/languages.hpp:155]:
```suggestion
    {Key::No_wave_made, "Es wurden keine Wellenformen erstellt!"},

```

File: src/languages.hpp [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    {Key::Wrong_file_warning, "Warnung vor falscher Datei! "},
```
Comment (NP199) [inline src/languages.hpp:156]:
```suggestion
    {Key::Wrong_file_warning, "Warnung vor falscher Datei!"},

```

File: src/languages.hpp [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+    {Key::Wrong_file_type, "Falscher Dateityp! Versuchen Sie es erneut!  "},
```
Comment (NP199) [inline src/languages.hpp:157]:
```suggestion
    {Key::Wrong_file_type, "Falscher Dateityp! Versuchen Sie es erneut!"},

```

File: src/languages.hpp [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+    {Key::Gn_trng_data_pop, "Trainingsdaten Generieren"},
+    {Key::Data_upload, "Daten hochladen"},
+    {Key::Upload_success, "Datei erfolgreich hochgeladen! "},
+    {Key::Upload_failure, "Das Hochladen der Datei ist fehlgeschlagen!  "},
```
Comment (NP199) [inline src/languages.hpp:161]:
```suggestion
    {Key::Upload_failure, "Das Hochladen der Datei ist fehlgeschlagen!"},

```

File: src/languages.hpp [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+    {Key::Devices_found, "Geräte Gefunden"},
+    {Key::Usr_curnt_wave, "Aktuelle Wellenform Des Benutzers"},
+    {Key::Wv_from_file, "Wellenform Aus Datei"},
+    {Key::WvForms_warning, "Warnung vor Wellenformen! "},
+    {Key::No_wave_made, "Es wurden keine Wellenformen erstellt! "},
+    {Key::Wrong_file_warning, "Warnung vor falscher Datei! "},
+    {Key::Wrong_file_type, "Falscher Dateityp! Versuchen Sie es erneut!  "},
+    {Key::Gn_trng_data_pop, "Trainingsdaten Generieren"},
+    {Key::Data_upload, "Daten hochladen"},
+    {Key::Upload_success, "Datei erfolgreich hochgeladen! "},
```
Comment (NP199) [inline src/languages.hpp:160]:
```suggestion
    {Key::Upload_success, "Datei erfolgreich hochgeladen!"},

```

File: src/main.cpp [hunk 12]
```diff
@@ -430,8 +422,10 @@ int main() {
     }
 
     // Generate training data Menu
+    static std::string api_message;
```
Comment (NP199) [inline src/main.cpp:425]:
```suggestion
    static std::string api_message{};

```

File: src/sendData.hpp [hunk 13]
```diff
@@ -1,43 +1,63 @@
...
+#ifndef SENDDATA_HEADER_HPP
+#define SENDDATA_HEADER_HPP
+
+#include "jasonhandler.hpp"
+#include <curl/curl.h>
+#include <fmt/format.h>
+
+inline std::string sendData(nlohmann::json const &config,
+                            const nlohmann::json &myJson) {
+  std::string api_message = "empty";
```
Comment (NP199) [inline src/sendData.hpp:12]:
```suggestion
  std::string api_message{"empty"};

```

File: src/sendData.hpp [hunk 14]
```diff
@@ -1,43 +1,63 @@
...
+
+  CURL *curl;
+  CURLcode res;
```
Comment (NP199) [inline src/sendData.hpp:15]: When is res needed?

File: src/sendData.hpp [hunk 15]
```diff
@@ -1,43 +1,63 @@
...
+
+  curl_global_init(CURL_GLOBAL_DEFAULT);
+  curl = curl_easy_init();
+
+  if (curl) {
+    std::string url = load_json<std::string>(config, "api", "url");
+
+    curl_mime *mime = curl_mime_init(curl);
+    curl_mimepart *part;
```
Comment (NP199) [inline src/sendData.hpp:24]:
```suggestion
    curl_mimepart *part = curl_mime_addpart(mime);

```

File: src/sendData.hpp [hunk 16]
```diff
@@ -1,43 +1,63 @@
...
+
+    part = curl_mime_addpart(mime);
```
Comment (NP199) [inline src/sendData.hpp:26]:
```suggestion


```

File: src/settingspopup.hpp
Comment (NP199) [inline src/settingspopup.hpp:1]: What are the changes to this file?

File: src/apihandler.hpp (hunk 2 above)
Comment (R-Abbasi) [reply src/apihandler.hpp:27]: Didn't change `apihandler.hpp` nor can I tell you of the exact reason why `git` found it modified when `push`ing the branch. Changed parts, as stated in the `pr` body, are: `generateTrainingData.cpp`, `saves_popup.cpp`, `languages.hpp`, `sendData.hpp` and `enum class` plus few statements in `SideBarMenu.cpp` to use the current `Key` for the required string literals.

File: src/generateTrainingData.cpp
[the remark this answers is not in the document]
Comment (R-Abbasi) [reply src/generateTrainingData.cpp:1]: My approach for the tasks is: develop on Windows in `Debug` config, then `Release` and finally on Linux using GCC. When the change is considerable in size, like now which is almost double for `generateTrainingData.cpp`, and it is sparsely spread across the whole file, I simply paste the written code from Windows to Linux. "Probably" (?) it's the reason GitHub hasn't shown the changes properly.

File: src/info_popup.cpp [hunk 17]
```diff
@@ -1,17 +1,15 @@
+#include <imgui.h>
```
[the remark this answers is not in the document]
Comment (R-Abbasi) [reply src/info_popup.cpp:1]: Because we have Dear ImGui stuff there.

File: src/languages.hpp [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+    {Key::Anlyz_crnt_waveform, "Aktuelle Wellenform analysieren"},
+    {Key::Gnrt_trning_data, "Trainingsdaten generieren"},
+    {Key::Timing_Belt, "Zahnriemen"},
+    {Key::Fuel_Delivery_Pump, "Kraftstoff-Förderpumpe"},
+    {Key::Common_Rail_Pressure, "Common-Rail-Druck"},
+    {Key::Help, "Hilfe"},
+    {Key::HelpLink, "Klicken Sie auf den Button um zur Hilfeseite zu gelangen"},
+    {Key::Dvc_search, "Suche Geräte"},
+    {Key::Save, "Speichern"},
+    {Key::Save_warning, "Speicherwarnung! "},
```
[the remark this answers is not in the document]
Comment (R-Abbasi) [reply src/languages.hpp:143]: That extra white space character added deliberately in some of the strings is basically to have a little longer widget.

File: src/main.cpp (hunk 12 above)
Comment (R-Abbasi) [reply src/main.cpp:425]: Not required that much, to me, since `std::string` has an explicit default constructor.

File: src/sendData.hpp (hunk 14 above)
Comment (R-Abbasi) [reply src/sendData.hpp:15]: Line 44:
```
// Sending the request
res = curl_easy_perform(curl);
```

**Pull request closed without merging**